### PR TITLE
Sparse CSR CUDA: Add block sparse support for torch.triangular_solve

### PR DESCRIPTION
**Pull request by IvanYashchuk**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #68711
* #68710
* __->__ #68709

This PR adds support for triangular solver with a block CSR matrix.

cc @nikitaved @pearu @cpuhrsch @IvanYashchuk @ngimel

Differential Revision: [D33066067](https://our.internmc.facebook.com/intern/diff/D33066067)

### CODE REVIEW

**Comment by IvanYashchuk** [comment]
ROCm failure is real. Strange that it builds locally and tests passed, need to verify that again. Maybe an older hipSPARSE version is used in CI.

File: aten/src/ATen/native/sparse/cuda/SparseBlasImpl.cpp [hunk 1]
```diff
@@ -109,6 +109,251 @@ void inline col_indices_and_values_resize_(const Tensor& input, int64_t nnz) {
...
+  //     prepare_dense_matrix_for_cusparse(B, X_->strides());
+
+  c10::MaybeOwned<Tensor> X_ = prepare_column_major_matrix_for_cusparse(X);
+  c10::MaybeOwned<Tensor> B_ = prepare_column_major_matrix_for_cusparse(B);
+
+  IntArrayRef X_strides = X_->strides();
+  auto ndim = X_->dim();
+  TORCH_INTERNAL_ASSERT_DEBUG_ONLY(ndim == 2);
+  bool is_X_row_major = (X_strides[ndim - 1] == 1);
+  // int ldx = is_X_row_major ? cuda_int_cast(X_strides[ndim - 2], "ldx")
```
Comment (cpuhrsch) [inline aten/src/ATen/native/sparse/cuda/SparseBlasImpl.cpp:264]: nit: Did you mean to remove this?

**Review by cpuhrsch — APPROVED** [review]

File: aten/src/ATen/native/sparse/cuda/SparseBlasImpl.cpp (hunk 1 above)
Comment (IvanYashchuk) [reply aten/src/ATen/native/sparse/cuda/SparseBlasImpl.cpp:264]: Yes, I need to clean up this and there is another commented code is left in a few places.

**Comment by cpuhrsch** [comment]
@cpuhrsch has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33066067).

**Comment by IvanYashchuk** [comment]
ROCm build fails https://ci.pytorch.org/jenkins/job/pytorch-builds/job/pytorch-linux-bionic-rocm4.3.1-py3.6-build/17076/

**Comment by IvanYashchuk** [comment]
It's legit. It was confusing to me because I tested previously locally and everything was building fine. The problem was that I tested with ROCm 4.5.0 and CI has older version 4.3.1. I will add skips for ROCm.

**Comment by cpuhrsch** [comment]
@cpuhrsch has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33066067).

**Pull request closed without merging**